### ISSUE TITLE
event-based activation (send button) is not working on window desktop outlook.

### ISSUE THREAD

**Issue by zjmchoi** [OfficeDev/office-js#3335]
> Message from office-js bot:
We’re closing this issue because it has been inactive for a long time. We’re doing this to keep the issues list manageable and useful for everyone. If this issue is still relevant for you, please create a new issue.
> Thank you for your understanding and continued feedback.

## Post how-to questions about your code on Microsoft Q&A

We have several forums where you can ask the developer community questions about your code. Try one of the following forums to ask technical questions about:
- [Developing Office Add-ins for Word, Excel, PowerPoint, Outlook, Project, Visio, and OneNote](https://docs.microsoft.com/answers/topics/office-addins-dev.html)
- [Access programming in Office](https://docs.microsoft.com/answers/topics/office-access-dev.html)
- [Developing COM or VSTO Add-ins for Office](https://docs.microsoft.com/answers/topics/office-vsto-com-dev.html)
- [Office JavaScript API](https://docs.microsoft.com/answers/topics/office-js-dev.html)
- [SharePoint and OneDrive development](https://docs.microsoft.com/answers/topics/sharepoint-dev.html)
- [Deploying and publishing Office or SharePoint apps to AppSource](https://docs.microsoft.com/answers/topics/microsoft-365-apps-publishing-dev.html)

For additional information about JavaScript programming, see the [JavaScript at Microsoft](https://docs.microsoft.com/javascript/) community hub.

**Comment by zjmchoi** [comment]
Currently, I am developing an Outlook Add-in.

1. Goal:
    - The goal is to work with same code in both Outlook Web (chrome, edge, safari) and Outlook Application (Mac, Window), except Outlook mobile..
    - Using the event based activation function and having a Microsoft 365 account.
2. Scenario: 
   - For testing, a simple function that when I schedule an event in the schedule window, and press the send button, I write "1234" in the mail body and close the window. (I am testing locally, not distributing.)
3. Issue: 
   - Currently, this function does not work only in Window desktop Outlook.
4. Version:
   - Outlook: microsoft 365 mso (version 2304 build 16.0.16327.20200) 32 bit
   - OS: Window 10 Pro (21H2, 190442.2846)
5. Code:
    function onItemSendHandler(event) {
      Office.context.mailbox.item.body.setAsync(
        'test1234',
        { asyncContext: event, coercionType: "html" },
        function (asyncResult) {
          let event = asyncResult.asyncContext;
          if (asyncResult.status === Office.AsyncResultStatus.Failed) {
            event.completed({ allowEvent: false });
            return;
          }
        });
    }
    Office.actions.associate("onAppointmentSendHandler", onItemSendHandler);
 6. Question:
   (1)  Why is it not working? (Are there any restrictions?)
   (2) To use event-based activation, I know that I need to deploy it through administrator, not through App Source.
        Do I need to apply this method when testing locally?

**Comment by ndeleuze** [comment]
@zjmchoi would you be able to share your manifest file?

**Comment by zjmchoi** [comment]
I send the manifest file you requested.
[manifest.xml.zip](https://github.com/OfficeDev/office-js/files/11512909/manifest.xml.zip)

**Comment by ndeleuze** [comment]
It looks like you need to also complete the event if the call to setAsync is successful.

Try something like:

`function onItemSendHandler(event) {
  console.log("onItemSendHandler Entered");
  Office.context.mailbox.item.body.setAsync(
    'test1234',
    { asyncContext: event, coercionType: "html" },
    function (asyncResult) {
      let event = asyncResult.asyncContext;
      if (asyncResult.status === Office.AsyncResultStatus.Failed) {
        event.completed({ allowEvent: false, errorMessage: "Unable to set appointment body."  });
        return;
      }
      console.log("onItemSendHandler allowEvent: true");
      event.completed({ allowEvent: true });
      return;
    }
  );
}`

You also need to ensure that you have the OnAppointmentSend event added to the LaunchEvent section of your manifest, like:

`<ExtensionPoint xsi:type="LaunchEvent">
        <LaunchEvents>
            <LaunchEvent Type="OnAppointmentSend" FunctionName="onAppointmentSendHandler" SendMode="PromptUser" />
        </LaunchEvents>
        <SourceLocation resid="rt0" />
</ExtensionPoint>`

**Comment by zjmchoi** [comment]
As you explained, I added code to complete the event if the call to setAsync is successful.
Then, the text ("test1234") was successfully written in the mail body.
thanks.
And, I did an additional test and it didn't work again.
1. my add-in saved the "access token" to local storage.
2, When on send, I tried to read it and write it in the mail body.
   -> A smart alert message is displayed. ("my add-in is taking longer than expected." )

The code is shown below.
-------------------------------------------------------------------------------------------------------------
function onItemSendHandler(event) {
  console.log("onItemSendHandler Entered");

  let accessToken = JSON.parse(JSON.parse(window.localStorage.getItem('persist:root')).auth).myAddIn_accessToken;

  Office.context.mailbox.item.body.setAsync(
    accessToken,
    {asyncContext: event, coercionType: "html"},
    function (asyncResult) {
      let event = asyncResult.asyncContext;
      if (asyncResult.status === Office.AsyncResultStatus.Failed) {
        event.completed({allowEvent: false, errorMessage: "Unable to set appointment body."});
        return;
      }
      event.completed({allowEvent: true});
      return;
    }
  ); 
}
Office.actions.associate("onAppointmentSendHandler", onItemSendHandler);
-------------------------------------------------------------------------------------------------------------

Can't read local storage in window?
If not, is there any other way?

**Comment by ndeleuze** [comment]
I would try to debug this one step at a time (i.e. pull apart the line where you get the accessToken and log each operation to ensure that this is working as expected). Usually this "my add-in is taking longer than expected" message is an indicator that some unhandled error occurred in your add-in.

**Comment by zjmchoi** [comment]
In the code above, I am not getting "accessToken".

let accessToken = JSON.parse(JSON.parse(window.localStorage.getItem('persist:root')).auth).myAddIn_accessToken; <-- here.

After searching several documents and internet, I found that I can't use "window.localStorage" in event-based activation code.
So, I tried using the API below as an alternative way.
1. "OfficeRuntime.Storage"
2. "Office.sessionData"
However, all have failed.
Are there any other alternative way?
In other words, how to get data saved in store of React?

**Comment by ndeleuze** [comment]
You can use the Office.context.roamingSettings.get(), Office.context.roamingSettings.set(), and Office.context.roamingSettings.saveAsync() to store and read local data. Please read the documenation for this interface here: https://learn.microsoft.com/en-us/javascript/api/outlook/office.roamingsettings?view=outlook-js-preview

**Comment by zjmchoi** [comment]
It works fine. 
Thanks.
